### PR TITLE
IStartupFilter refactoring

### PR DESCRIPTION
**Pull request by kimbell**
* mapped endpoints controlled via IServiceCollection
* removed majority of IStartupFilters
* AllActuatorsStartupFilter is the only one setting up endpoints

Addresses #497

The main purpose of this change is to control what endpoints get mapped based on what has been registered in the `IServiceCollection`. This allows `MapAllActutators` to be the single mapping function; the others have been retained for binary compatibility, but they all end up in `MapAllActuators`. The code shouldn't care if the client calls both `AddEnvActuator` and `AddAllActuators`. It has code that avoids duplicate sets of endpoints.

We are now left with one `IStartupFilter` that sets up endpoints; `AllActuatorsStartupFilter`. All the others have been removed. The other `IStartupFilters` that are remaining deal with other things than endpoints. i.e. security.

This change can be very beneficial when it comes to setting up tests. Changing the contents of the `IServiceCollection` per test is easy. Changing the Http pipeline per test is trickier. This change then allows testing code to use the same Http pipeline, but based on what has been registered in `IServiceCollection`, each test can be focused.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/SteeltoeOSS/Steeltoe?pullRequest=543) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/SteeltoeOSS/Steeltoe?pullRequest=543) before we can accept your contribution.<br/><br/>:x: kimbell [sign now](https://cla.dotnetfoundation.org/SteeltoeOSS/Steeltoe?pullRequest=543)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/SteeltoeOSS/Steeltoe?pullRequest=543) it.</sub>

File: src/Management/src/EndpointBase/Health/HealthConverter.cs [hunk 1]
```diff
@@ -26,17 +26,10 @@ public override void Write(Utf8JsonWriter writer, HealthEndpointResponse value,
                     writer.WriteString("description", health.Description);
                 }
 
-                if (health.Details != null && health.Details.Count > 0)
+                foreach (var detail in health.Details)
```
Comment (TimHess) [inline src/Management/src/EndpointBase/Health/HealthConverter.cs:29]: This looks like a reversal of #518, maybe slipped through during a rebase?

Comment (kimbell) [reply src/Management/src/EndpointBase/Health/HealthConverter.cs]: I was at git hell level 1 and tried to be 'clever' and ended up in git hell level 2. 
I've made some changes that I hope will resolve things.

**Comment by jkonicki** [comment]
See #503 for more information

File: src/Management/src/EndpointCore/Extensions/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+﻿using Microsoft.AspNetCore.Builder;
```
Comment (TimHess) [inline src/Management/src/EndpointCore/Extensions/ServiceCollectionExtensions.cs:1]:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the Apache 2.0 License.
// See the LICENSE file in the project root for more information.

﻿﻿using Microsoft.AspNetCore.Builder;
```

File: src/Management/src/EndpointCore/Internal/EndpointEntry.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using Microsoft.AspNetCore.Routing;
```
Comment (TimHess) [inline src/Management/src/EndpointCore/Internal/EndpointEntry.cs:1]:
```suggestion
﻿// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the Apache 2.0 License.
// See the LICENSE file in the project root for more information.

using Microsoft.AspNetCore.Routing;
```

File: src/Management/src/EndpointCore/Internal/EndpointTracking.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
```
Comment (TimHess) [inline src/Management/src/EndpointCore/Internal/EndpointTracking.cs:1]:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the Apache 2.0 License.
// See the LICENSE file in the project root for more information.

﻿﻿using System.Collections.Generic;
```

File: src/Management/src/EndpointCore/Extensions/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+﻿using Microsoft.AspNetCore.Builder;
```
Comment (TimHess) [inline src/Management/src/EndpointCore/Extensions/ServiceCollectionExtensions.cs:5]:
```suggestion
using Microsoft.AspNetCore.Builder;
```

**Comment by kimbell** [comment]
I'm confused. When looking at https://codecov.io/gh/SteeltoeOSS/Steeltoe/pull/543/changes it looks like the majority of reduced code coverage is in files that I haven't touched; CircuitBreaker and Messaging

**Comment by hananiel** [comment]
@kimbell thank you so much for looking at this and contributing. Sorry for taking so long to respond, we are a little busy with our next release. I will work with you to get some of these changes merged in.  This PR addresses a number of issues and I think it will be better to address them one at a time:
1. _Adding multiple route endpoints to the code will cause errors when multiple paths match_. Although this behavior is expected it's not ideal. To solve this, in [MapActuatorEndpoint ](https://github.com/SteeltoeOSS/Steeltoe/blob/835cfd21ea1bd2cb8a413c8fb406dd596eedd35f/src/Management/src/EndpointCore/ActuatorRouteBuilderExtensions.cs#L131) , we can add something like this: 

```csharp
if (!endpoints.DataSources.Any(d => d.Endpoints.Any(ep => ((RouteEndpoint)ep).RoutePattern.RawText == pattern.RawText)))
{   
   // Only then call MapMethods ... 
```
  I did a  quick check but ... something along these lines. Basically using routing information itself to make the calls idempotent.

2. _Adding multiple IStartup filters is not a best practice._ Though this is true in general, in our case we don't have any ordering issues, since we are only using this for routes. And then the middlewares arent invoked unless the pattern matches. Granted they are still registered with router middleware multiple times but with the above change, we remedy that. 

3. _Testing_ - this is an area for improvement as well as reducing complexity. I feel the PR itself had a lot of redirection and was having some difficulty following it. And I am not still sure how you handled the platform-specific code for Heapdump and thread dump. 

My suggestion is to address these issues one at a time - starting with multiple AddXXXActuator should not result in a routing error that we see.  Also, the above code was just a suggestion - if you find a different way - that's totally fine. and more test coverage for the changes, please.

**Comment by kimbell** [comment]
@hananiel your suggestion for detecting duplicate routes is not something I thought about, but should work just as well. Allows us to remove the `EndpointTracking` object I introduced.

As for multiple `IStartupFilters`, I have no problem with having more than one. The question is what they do. I asked a question on the ASP.NET Core repo and was told calling `UseEndpoints` multiple times is not ideal. If there is a simple way of avoiding that, it should be done. https://github.com/dotnet/aspnetcore/discussions/27503

I don't see how any fix in `MapAcutatorEndpoint` will solve that since it gets called inside the `UseEndpoints`.

In my PR, we now only have one `IStartupFilter` that configures routes. The removal of the others is a potential breaking change if someone has based their code on these startup filters. We could add them back in again and mark them obsolete. They don't need to perform any code. 

As for the platform specific code, the idea is to leave that to the `IServiceCollection` setup. This contains platform checks and adds `EndpointEntry` instances to the `IServiceCollection`. When we get to the Map step, it enumerates the `EndpointEntry` that have been registered. 

Here I noticed that there are differences between heapdump and threaddump that I haven't spotted.

https://github.com/SteeltoeOSS/Steeltoe/blob/master/src/Management/src/EndpointCore/HeapDump/EndpointServiceCollectionExtensions.cs#L40
 ```cs
if (IsHeapDumpSupported())
            {
                services.AddActuatorManagementOptions(config);

                var options = new HeapDumpEndpointOptions(config);
                services.TryAddSingleton<IHeapDumpOptions>(options);
                services.RegisterEndpointOptions(options);
```

https://github.com/SteeltoeOSS/Steeltoe/blob/master/src/Management/src/EndpointCore/ThreadDump/EndpointServiceCollectionExtensions.cs does not contain a `Platform.IsWindows` check; this is only performed in `MapAllActuators`.

I guess this is a bug in the current code. If clients don't use `MapAllActuators`, it's possible to set up the ThreadDump endpoint on non-Windows machines.

**Comment by hananiel** [comment]
> As for multiple `IStartupFilters`, I have no problem with having more than one. The question is what they do. I asked a question on the ASP.NET Core repo and was told calling `UseEndpoints` multiple times is not ideal. If there is a simple way of avoiding that, it should be done. [dotnet/aspnetcore#27503](https://github.com/dotnet/aspnetcore/discussions/27503)
> 
  I know its not ideal, but I don't feel its all that bad either. All it's doing is at calling multiple "configure" methods. Which if the configure method is doing nothing, it's just a small cost. 

> I don't see how any fix in `MapAcutatorEndpoint` will solve that since it gets called inside the `UseEndpoints`.
Adding IStartupFilter allows the framework to call StartupFilter.Configure ->  UseEndpoints -> MapActuatorEndpoint. 

> In my PR, we now only have one `IStartupFilter` that configures routes. The removal of the others is a potential breaking change if someone has based their code on these startup filters. We could add them back in again and mark them obsolete. They don't need to perform any code.

That is fine with me, but let me discuss, assess the impact of that decision and get back to you.

**Comment by kimbell** [comment]
@hananiel I created a new branch for a 'take 2' on this area focusing on point 1 in your list. 

I haven't created a PR yet, since I'm a bit unsure what issue to link it to. I feel the changes I've made don't directly relate to the original #497. Should we create a new issue?

I have a diff that shows the changes so far. Are these more in line with your thinking?
https://github.com/SteeltoeOSS/Steeltoe/compare/master...kimbell:startupFilter_take2

**Comment by jkonicki** [comment]
Moving out of 3.0.2.  We will look at this work for next release.

**Comment by hananiel** [comment]
> @hananiel I created a new branch for a 'take 2' on this area focusing on point 1 in your list.
> 
> I haven't created a PR yet, since I'm a bit unsure what issue to link it to. I feel the changes I've made don't directly relate to the original #497. Should we create a new issue?
> 
> I have a diff that shows the changes so far. Are these more in line with your thinking?
> [master...kimbell:startupFilter_take2](https://github.com/SteeltoeOSS/Steeltoe/compare/master...kimbell:startupFilter_take2)

No, I think we should just resolve the single issue of multiple registrations of middleware with as little code as possible. Lets just start with a fresh PR for that issue if you would like. This has been sitting for a while and is getting confusing.

**Comment by TimHess** [comment]
This PR is being replaced by #554

**Pull request closed without merging**